rfcs/0100-sparestir-dialect.md

```diff
@@ -0,0 +1,658 @@
+
+We will outline the detailed design in the next section.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This section outlines the design of SparseTIR, and its interaction with existing components in TVM.
+
+## D0: Programming Interface
+
+A generic SparseTIR program looks like the following, the workload is Sampled-Dense-Dense-Matrix-Multiplication (SDDMM):
+
+```python
+@T.prim_func
+def sddmm(
+    a: T.handle,
+    b: T.handle,
+    x: T.handle,
+    y: T.handle,
+    indptr: T.handle,
+    indices: T.handle,
+    m: T.int32,
+    n: T.int32,
+    feat_size: T.int32,
+    nnz: T.int32,
+) -> None:
+    T.func_attr({"global_symbol": "main", "tir.noalias": True, "sparse_tir_level": 2})
+    # sparse axes
+    I = T.dense_fixed(m)
+    J = T.sparse_variable(I, (n, nnz), (indptr, indices), "int32")
+    J_detach = T.dense_fixed(n)
+    K = T.dense_fixed(feat_size)
+    # sparse buffers
+    A = T.match_sparse_buffer(a, (I, K), "float32")
+    B = T.match_sparse_buffer(b, (J_detach, K), "float32")
+    X = T.match_sparse_buffer(x, (I, J), "float32")
+    Y = T.match_sparse_buffer(y, (I, J), "float32")
+    # sparse iterations
+    with T.sp_iter([I, J, K], "SSR", "sddmm") as [i, j, k]:
+        with T.init():
+            Y[i, j] = 0.0
+        Y[i, j] = Y[i, j] + A[i, k] * B[j, k] * X[i, j]
+```
+
+where we have constructs like **sparse axes**,  **sparse buffers** and **sparse iterations**.
+
+### Sparse Axis
+Sparse axis is a generation of per-dimensional level formats in TACO where we annotate each dimension of a format as **dense**/**sparse** (this dimension is stored in dense or compressed storage) and **fixed**/**variable** (this dimension's extent is fixed or variable). For **sparse**/**variable** axes, we need to specify its dependent axis.
+
+- For axes that are **sparse**, we need to specify a `indices` array to store the column indices.
+- For axes that are **variable**, we need to specify an `indptr` (short for indices pointer) array to store the start offset of each row because the row length is variable and we cannot simply compute element offset with an affine map of indices.
+- An axes that is both **sparse** and **variable** need to be specified with both **indices** and **indptr** array.
+
+```python
+I = T.dense_fixed(m)
+# J1 is a sparse fixed axis, whose dependent axis is I
+# it has maximum length n and number of non-zero elements per row: c,
+# the column indices data are stored in the region started from indices_1 handle,
+# and the index data type (in indices array) is int32.
+J1 = T.sparse_fixed(I, (n, c), indices_1, idtype="int32")
+# J2 is a dense variable axis, whose dependent axis is I,
+# it has a maximum length of n,
+# the indptr data are stored in the region started from indptr_2 handle,
+# and the index data type (in indptr array) is int32.
+J2 = T.dense_variable(I, n, indptr_2, idtype="int32")
```

Comment:
Thinking on it, it would also allow a straightforward way to express ragged buffers where the extent of an axis depends on the axis `I`, but doesn't depend on an external buffer.  In these cases, it could be a `PrimExpr` that depends on `I` (e.g. a triangular buffer `extent = I`), and doesn't need to be based on a `indptr` buffer.

Reply:
Here the `n` refers `maximum` length, which means the `extent` if we view it as a dense matrix. Such information can help us generate some hints (e.g. the value inside indices buffer is less or equal to `n`), and in SparseTIR we have an `assume_buffer_domain` to describe it.

Reply:
Ah, thank you.  I'd been having some confusion between "extent" of the sparse tensor and "extent" of the dense backing buffer.